File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueEntropiaAproximada.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Entropia aproximada (ApEn), inspirado no NIST SP800-22. Mede a
 * previsibilidade local: para uma sequência aleatória, a chance de repetir um
 * bloco de m bits deve cair suavemente conforme m cresce. Estruturas
 * periódicas/recorrentes produzem ApEn anômala.
 *
 * O estatístico chi2 = 2n(ln2 - ApEn) tem viés para n finito, então NÃO usamos
 * um valor esperado teórico. Comparamos o keystream com um CONTROLE de
 * random_bytes nas MESMAS condições - se a cifra for boa, os dois devem ser
 * estatisticamente indistinguíveis.
 */
public class AtaqueEntropiaAproximada implements AtaqueInterface {
    private final int tamanho;
    private final int m;
    private final int controles;
    private final int amostrasKeystream;
    private final double limiteZ;

    public AtaqueEntropiaAproximada() {
        this(4096, 8, 12, 3, 5.0);
    }

    public AtaqueEntropiaAproximada(
            int tamanho, int m, int controles, int amostrasKeystream, double limiteZ) {
        this.tamanho = tamanho;
        this.m = m;
        this.controles = controles;
        this.amostrasKeystream = amostrasKeystream;
        this.limiteZ = limiteZ;
    }

    @Override
    public String nome() {
        return "Entropia aproximada (ApEn vs controle aleatório)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] chave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8);
        byte[] primeiro = alvo.gerarKeystreamBruto(
                chave, Util.randomBytes(alvo.tamanhoIv()), "enc", 64);
        if (primeiro == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        double[] chiControle = new double[this.controles];
        for (int c = 0; c < this.controles; c++) {
            chiControle[c] = this.estatistico(Util.randomBytes(this.tamanho));
        }
        double mediaControle = 0.0;
        for (double v : chiControle) {
            mediaControle += v;
        }
        mediaControle /= chiControle.length;
        double variancia = 0.0;
        for (double v : chiControle) {
            double d = v - mediaControle;
            variancia += d * d;
        }
        double desvioControle = Math.sqrt(variancia / Math.max(1, chiControle.length - 1));

        double[] chiKeystream = new double[this.amostrasKeystream];
        for (int k = 0; k < this.amostrasKeystream; k++) {
            byte[] ks = alvo.gerarKeystreamBruto(
                    chave, Util.randomBytes(alvo.tamanhoIv()), "enc", this.tamanho);
            chiKeystream[k] = this.estatistico(ks);
        }
        double mediaKeystream = 0.0;
        for (double v : chiKeystream) {
            mediaKeystream += v;
        }
        mediaKeystream /= chiKeystream.length;

        double z = desvioControle > 0 ? (mediaKeystream - mediaControle) / desvioControle : 0.0;
        boolean vulneravel = Math.abs(z) > this.limiteZ;

        Map<String, Object> dados = new HashMap<>();
        dados.put("chi_keystream", mediaKeystream);
        dados.put("chi_controle", mediaControle);
        dados.put("z", z);

        return new ResultadoAtaque(
                nome(),
                vulneravel,
                vulneravel ? Severidade.MEDIA : Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "chi2 keystream=%.1f, controle=%.1f (sd=%.1f), z=%.2f (limite=%.1f)",
                        mediaKeystream,
                        mediaControle,
                        desvioControle,
                        z,
                        this.limiteZ),
                dados);
    }

    /** chi2 = 2n(ln2 - ApEn(m)), com ApEn = phi(m) - phi(m+1). */
    private double estatistico(byte[] bytes) {
        int[] bits = this.paraBits(bytes);
        int n = bits.length;
        double apen = this.phi(bits, this.m, n) - this.phi(bits, this.m + 1, n);
        return 2 * n * (Math.log(2) - apen);
    }

    private double phi(int[] bits, int m, int n) {
        int total = 1 << m;
        int[] contagem = new int[total];
        int janelas = n - m + 1;

        for (int i = 0; i < janelas; i++) {
            int v = 0;
            for (int j = 0; j < m; j++) {
                v = (v << 1) | bits[i + j];
            }
            contagem[v]++;
        }

        double soma = 0.0;
        for (int c : contagem) {
            if (c > 0) {
                double p = (double) c / janelas;
                soma += p * Math.log(p);
            }
        }

        return soma;
    }

    private int[] paraBits(byte[] bytes) {
        int[] bits = new int[bytes.length * 8];
        int idx = 0;
        for (int i = 0; i < bytes.length; i++) {
            int b = bytes[i] & 0xFF;
            for (int j = 7; j >= 0; j--) {
                bits[idx++] = (b >> j) & 1;
            }
        }
        return bits;
    }
}

File: packages/go/internal/seguranca/ataques/ataque_entropia_aproximada.go
package ataques

import (
	"fmt"
	"math"

	"criptografia/internal/seguranca"
)

// OpcoesEntropiaAproximada configura o AtaqueEntropiaAproximada.
type OpcoesEntropiaAproximada struct {
	Tamanho           int
	M                 int
	Controles         int
	AmostrasKeystream int
	LimiteZ           float64
}

// AtaqueEntropiaAproximada mede a entropia aproximada (ApEn) do keystream e a
// compara com um controle de bytes aleatórios nas mesmas condições.
type AtaqueEntropiaAproximada struct {
	tamanho           int
	m                 int
	controles         int
	amostrasKeystream int
	limiteZ           float64
}

// NovoAtaqueEntropiaAproximada cria o ataque. Sem argumento, usa tamanho = 4096,
// m = 8, controles = 12, amostrasKeystream = 3 e limiteZ = 5.0.
func NovoAtaqueEntropiaAproximada(opcoes ...OpcoesEntropiaAproximada) *AtaqueEntropiaAproximada {
	o := OpcoesEntropiaAproximada{Tamanho: 4096, M: 8, Controles: 12, AmostrasKeystream: 3, LimiteZ: 5.0}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueEntropiaAproximada{
		tamanho:           o.Tamanho,
		m:                 o.M,
		controles:         o.Controles,
		amostrasKeystream: o.AmostrasKeystream,
		limiteZ:           o.LimiteZ,
	}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueEntropiaAproximada) Nome() string {
	return "Entropia aproximada (ApEn vs controle aleatório)"
}

// Executar roda o ataque.
func (a *AtaqueEntropiaAproximada) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	chiControle := make([]float64, 0, a.controles)
	for c := 0; c < a.controles; c++ {
		chiControle = append(chiControle, a.estatistico(seguranca.RandomBytes(a.tamanho)))
	}
	mediaControle := 0.0
	for _, v := range chiControle {
		mediaControle += v
	}
	mediaControle /= float64(len(chiControle))
	variancia := 0.0
	for _, v := range chiControle {
		variancia += (v - mediaControle) * (v - mediaControle)
	}
	desvioControle := math.Sqrt(variancia / math.Max(1, float64(len(chiControle)-1)))

	chiKeystream := make([]float64, 0, a.amostrasKeystream)
	for k := 0; k < a.amostrasKeystream; k++ {
		ks := alvo.GerarKeystreamBruto([]byte(alvo.ChaveDeTeste()), seguranca.RandomBytes(alvo.TamanhoIv()), "enc", a.tamanho)
		chiKeystream = append(chiKeystream, a.estatistico(ks))
	}
	mediaKeystream := 0.0
	for _, v := range chiKeystream {
		mediaKeystream += v
	}
	mediaKeystream /= float64(len(chiKeystream))

	z := 0.0
	if desvioControle > 0 {
		z = (mediaKeystream - mediaControle) / desvioControle
	}
	vulneravel := math.Abs(z) > a.limiteZ

	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevMedia
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"chi2 keystream=%.1f, controle=%.1f (sd=%.1f), z=%.2f (limite=%.1f)",
			mediaKeystream, mediaControle, desvioControle, z, a.limiteZ,
		),
		Dados: map[string]any{
			"chi_keystream": mediaKeystream,
			"chi_controle":  mediaControle,
			"z":             z,
		},
	}, nil
}

// estatistico calcula chi2 = 2n(ln2 - ApEn(m)), com ApEn = phi(m) - phi(m+1).
func (a *AtaqueEntropiaAproximada) estatistico(bytes []byte) float64 {
	bits := a.paraBits(bytes)
	n := len(bits)
	apen := a.phi(bits, a.m, n) - a.phi(bits, a.m+1, n)
	return 2 * float64(n) * (math.Ln2 - apen)
}

// phi calcula a entropia aproximada para blocos de m bits.
func (a *AtaqueEntropiaAproximada) phi(bits []int, m, n int) float64 {
	total := 1 << m
	contagem := make([]int, total)
	janelas := n - m + 1

	for i := 0; i < janelas; i++ {
		v := 0
		for j := 0; j < m; j++ {
			v = (v << 1) | bits[i+j]
		}
		contagem[v]++
	}

	soma := 0.0
	for _, c := range contagem {
		if c > 0 {
			p := float64(c) / float64(janelas)
			soma += p * math.Log(p)
		}
	}

	return soma
}

// paraBits converte bytes em bits, do mais significativo ao menos.
func (a *AtaqueEntropiaAproximada) paraBits(bytes []byte) []int {
	bits := make([]int, 0, len(bytes)*8)
	for i := 0; i < len(bytes); i++ {
		by := bytes[i]
		for b := 7; b >= 0; b-- {
			bits = append(bits, int((by>>b)&1))
		}
	}
	return bits
}
